icicle-compiler/test/Icicle/Test/Foreign/Array.hs

```diff
@@ -0,0 +1,404 @@
+{-# LANGUAGE NoImplicitPrelude #-}
+{-# LANGUAGE TemplateHaskell   #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE TupleSections     #-}
+{-# LANGUAGE ViewPatterns      #-}
+{-# LANGUAGE BangPatterns      #-}
+
+{-# OPTIONS_GHC -fno-warn-missing-signatures #-}
+
+module Icicle.Test.Foreign.Array where
+
+import           Control.Monad.IO.Class (liftIO)
+
+import qualified Data.List as List
+
+import           Foreign
+import           Foreign.LibFFI (RetType)
+import           Foreign.LibFFI.Base (mkStorableRetType)
+import           Foreign.LibFFI.FFITypes (ffi_type_uint64)
+import           Foreign.LibFFI.Types as X
+
+import           System.IO
+
+import           Test.QuickCheck
+import           Test.QuickCheck.Property
+
+import qualified Prelude as Savage
+
+import           Disorder.Corpus
+import           Disorder.Core.IO
+
+import           P
+
+import           X.Control.Monad.Trans.Either
+
+import           Jetski
+
+import qualified Icicle.Common.Base as Icicle
+import           Icicle.Data.Time
+import           Icicle.Internal.Pretty
+import           Icicle.Test.Sea.Utils
+import           Icicle.Test.Arbitrary.Base
+import           Icicle.Test.Arbitrary
+
+
+
+-- array_eq (x, x) == true
+prop_array_eq_refl input
+  = testIO
+  $ runProp input "eq"
+  $ \ptr -> [ argPtr ptr, argPtr ptr]
+
+-- array_eq (x, y) == array_eq (y, x)
+prop_array_eq_comm input1 input2
+  = testIO
+  $ runWith input1
+  $ \ptr1
+  -> runProp input2 "eq_comm"
+  $ \ptr2
+  -> [ argPtr (wordPtrToPtr ptr1), argPtr ptr2 ]
+
+-- array_copy (x) == x
+prop_array_copy_eq input
+  = testIO
+  $ runProp input "copy_eq"
+  $ \ptr -> [ argPtr ptr ]
+
+-- y = array_copy (x)
+-- z = array_put_mutable (x, i, a)
+-- array_eq (x, y) == false
+prop_array_put_mutable_diff
+  = forAll arbitrary $ \input -> not (univalue (inputType input)) ==>
+    forAll arbitrary $ \i -> i >= 0 && i < length (inputArr input) ==>
+    forAll (genValForType (inputType input)) $ \v -> putTestCond input i v ==> testIO $ do
+      (vptr, vfree) <- pokeVal v
```

Comment:
it probably doesn't matter too much because it's a test, but I could imagine if you were running these in the repl trying to fix a problem, it might add up eventually